### ISSUE TITLE
Separating message format from encoding discussion, adding hard justifications

### ISSUE THREAD

**Issue by tschudin** [ssb-ngi-pointer/bendy-butt-spec#18]
Quote:

> Why a new feed format, when we already have gabby grove and bamboo?
>> Because we would want to offer a feed format that is simple and easy to implement in any programming language by reusing an established format (bencode) with existing encoder libraries.

There are so many "encoding religious wars" that it seems wise to separate this at the text level from the discussion of the (logical data structure) format needed for new feeds, instead of glossing over it in one sentence.

I note that the CBOR standard easily fulfills and probably exceeds the claims made for bencode ("established", "existing libraries"), so more arguments would be needed here, because bencode has shortcomings for sure.

Adding with SSB-BFE yet another binary encoding (instead of re-using CBOR or bencode for the structured ID data) also needs some hard justification - especially as falling back to fixed-width fields is exactly what above encodings wanted to overcome ...

**Comment by cryptix** [comment]
Thanks for bringing this up. First and foremost, I agree that this need better justifications.

In the interest of brevity I will only add three bits of metadata that lead to the decision to use bencode:

* The initial draft intended to use [bipf](https://github.com/ssbc/bipf) which I vetoed on the grounds of not mixing storage and network layers unnecessarily and that we shouldn't require people to implement a homegrown encoding format, too.

* I suggested to try bencode after having received feedback from folks that [gabbygrove](/?https://github.io/ssbc/ssb-spec-drafts/blob/master/drafts/draft-ssb-core-gabbygrove/00/draft-ssb-core-gabbygrove-00.html) (just including the link for people following along, I'm sure you know it and that it uses CBOR) isn't as easy to implement as I hoped for because libraries don't include the _canonical_ mode. The suggestion to use bencode is grounded in it's canonical purity since there is no off-switch for it and that it is older, less complex and thus more available.

* Another justification is that the intention was to _only_ use this format for metafeed management related tasks where all the message types and the data inside them are known and thus can reason about the short-communings of bencode and weather we can work around them, like with the BFE annotations. I personally want to stress that it's not intended as a general purpose feed format in the sense that the existing format can be use for free form data publishing.

**Comment by staltz** [comment]
I agree with cryptix and want to add that BFE has SSB-specific encodings for identifiers, like Message IDs, Feed IDs etc. These used to be algorithm-tagged, like `@*.ed25519` but now they are purpose-tagged, like `@*.bbfeed-v1`. COSE wouldn't come with that out of the box (I think), and would still require implementing such SSB-specific binary encoding of IDs.

I don't know the specifics too well, but I believe the purpose-tagged IDs are important for how meta feeds work.

Another aspect to this is pragmatism in our development work. We can endlessly discuss what is the ideal, but at some point we have to make a decision, and there will always be tradeoffs. We have already made a decision for BFE / bencode / bendy butt and going back on that now would mean wasted work. Sure, we might be sacrificing the pursuit of the *ideal scenario*, but what we have now is not bad at all and not hard to implement from scratch.
